controller/object_store_controller.go

```diff
@@ -0,0 +1,870 @@
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+var (
+	ErrObjectStorePVCNotReady        = errors.New("PVC not ready")
+	ErrObjectStoreVolumeNotReady     = errors.New("Volume not ready")
+	ErrObjectStorePVNotReady         = errors.New("PV not ready")
+	ErrObjectStoreDeploymentNotReady = errors.New("Deployment not ready")
+	ErrObjectStoreServiceNotReady    = errors.New("Service not ready")
+	ErrObjectStoreIngressNotReady    = errors.New("Ingress not ready")
+)
+
+type ObjectStoreController struct {
+	*baseController
+
+	namespace string
+	ds        *datastore.DataStore
+	s3gwImage string
+	uiImage   string
+
+	cacheSyncs []cache.InformerSynced
+}
+
+func NewObjectStoreController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string,
+	objectStoreImage string,
+	objectStoreUIImage string,
+) *ObjectStoreController {
+	osc := &ObjectStoreController{
+		baseController: newBaseController("object-store", logger),
+		namespace:      namespace,
+		ds:             ds,
+		s3gwImage:      objectStoreImage,
+		uiImage:        objectStoreUIImage,
+	}
+
+	ds.ObjectStoreInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    osc.enqueueObjectStore,
+			UpdateFunc: func(old, cur interface{}) { osc.enqueueObjectStore(cur) },
+			DeleteFunc: osc.enqueueObjectStore,
+		},
+	)
+	osc.cacheSyncs = append(osc.cacheSyncs, ds.ObjectStoreInformer.HasSynced)
+
+	return osc
+}
+
+func (osc *ObjectStoreController) Run(workers int, stopCh <-chan struct{}) {
+	osc.logger.Info("Starting Longhorn Object Store Controller")
+	defer osc.logger.Info("Shut down Longhorn Object Store Controller")
+	defer osc.queue.ShutDown()
+
+	if !cache.WaitForNamedCacheSync("longhorn object stores", stopCh, osc.cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(osc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (osc *ObjectStoreController) worker() {
+	for osc.processNextWorkItem() {
+	}
+}
+
+func (osc *ObjectStoreController) processNextWorkItem() bool {
+	key, quit := osc.queue.Get()
+	if quit {
+		return false
+	}
+	defer osc.queue.Done(key)
+
+	err := osc.syncObjectStore(key.(string))
+	if err == nil {
+		osc.queue.Forget(key)
+	} else if osc.queue.NumRequeues(key) < maxRetries {
+		osc.logger.WithError(err).Errorf("error syncing object store: \"%v\", retrying", err)
+		osc.queue.AddRateLimited(key)
+	} else {
+		utilruntime.HandleError(err)
+		osc.logger.WithError(err).Errorf("error syncing object store: \"%v\", giving up", err)
+		osc.queue.Forget(key)
+	}
```

Comment:
Since disabling the resyncPreiod is the recommeneded approach, we should not drop the object out of the queue because it will never be requeued 
```suggestion
	if err == nil {
		osc.queue.Forget(key)
		return
	} 
	
	osc.logger.WithError(err).Errorf("Failed to sync Longhorn ObjectStore: \"%v\"", key)
	osc.queue.AddRateLimited(key)

```

Reply:
We don't have to worry about the controller retrying a lot and wasting resources because we are adding it back to the queue in a `AddRateLimited` fashion. `AddRateLimited` will slow it down to 15 mins/30 mins so no worries about the resource consumption.